### PR TITLE
Subscription & Publication Data Model

### PR DESCRIPTION
**Pull request by helloeddiepo**

### CODE REVIEW

**Comment by helloeddiepo** [comment]
First iteration: creation of Publication data model + update existing Subscription data model. Prereqs to issues #43 & #44.

File: packages/fan-fix-data-model/PublicationModel.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { IdentifiableReference, List } from "@not-the-droids/exco-data-model";
+import { Inject } from "@not-the-droids/exco-ts-inject";
+import { Publication } from "./Publication";
+
+export type UpdatePublicationParams = Publication & {
+  categoryIds: ReadonlyArray<string>,
+};
```
Comment (maryseo070) [inline packages/fan-fix-data-model/PublicationModel.ts:7]: so UpdatePublicationParams has required properties creatorRef and categoryRefs? you may want to omit those.

File: packages/fan-fix-data-model/SubscriptionModel.ts [hunk 2]
```diff
@@ -2,9 +2,18 @@ import { IdentifiableReference, List } from "@not-the-droids/exco-data-model";
 import { Inject } from "@not-the-droids/exco-ts-inject";
 import { Subscription } from "./Subscription";
 
+export type UpdateSubscriptionParams = Subscription & {
+  appUserId: string,
+  publicationId: string,
+};
```
Comment (maryseo070) [inline packages/fan-fix-data-model/SubscriptionModel.ts:8]: omit refs for this type as well

File: packages/fan-fix-data-model/firebase/FirebasePublicationModel.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  FirestoreInjectable,
+  UnifiedFirestoreTypes,
+} from "@not-the-droids/exco-data-model-firebase";
+import { Inject, singletonInject } from "@not-the-droids/exco-ts-inject";
+import { CategoryModel } from "CategoryModel";
+import { AppUserModel } from "../AppUserModel";
+import { FirebaseAppUserModel } from "./FirebaseAppUserModel";
+import { FirebaseCategoryModel } from "./FirebaseCategoryModel";
+import { Publication } from "../Publication";
+import { CreatePublicationParams, PublicationModel, UpdatePublicationParams } from "../PublicationModel";
```
Comment (maryseo070) [inline packages/fan-fix-data-model/firebase/FirebasePublicationModel.ts:13]: can you sort these? 
also the Category Model has no path so will throw an error

File: packages/fan-fix-data-model/firebase/FirebasePublicationModel.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  const data: any = {};
+  if (categoryIds) {
+    data.categoryIds = categoryIds;
+  }
+  if (description) {
+    data.description = description;
+  }
+  if (price) {
+    data.price = price;
+  }
```
Comment (maryseo070) [inline packages/fan-fix-data-model/firebase/FirebasePublicationModel.ts:26]: you dont need conditions on required params

File: packages/fan-fix-data-model/firebase/FirebasePublicationModel.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  };
+
+  public getPublicationRef(id: string): IdentifiableReference<Publication> {
+    return this.publicationDataManager.getRef(id);
+  }
+
+  public getPublicationListByCreator(creatorId: string): List<Publication> {
+    return new FilteredList(this.publicationDataManager.list, publication => publication.creatorRef.id === creatorId)
+  }
```
Comment (maryseo070) [inline packages/fan-fix-data-model/firebase/FirebasePublicationModel.ts:74]: should be publication.creatorId = creatorId 
there will not be references in the database

File: packages/fan-fix-data-model/firebase/FirebaseSubscriptionModel.ts [hunk 6]
```diff
@@ -1,23 +1,40 @@
...
+import { FirebaseDataManager, FirestoreInjectable, UnifiedFirestoreTypes } from "@not-the-droids/exco-data-model-firebase";
 import { Inject, singletonInject } from "@not-the-droids/exco-ts-inject";
-import { CategoryModel } from "CategoryModel";
 import { AppUserModel } from "../AppUserModel";
-import { Subscription } from "../Subscription";
-import { SubscriptionModel } from "../SubscriptionModel";
 import { FirebaseAppUserModel } from "./FirebaseAppUserModel";
-import { FirebaseCategoryModel } from "./FirebaseCategoryModel";
+import { FirebasePublicationModel } from "./FirebasePublicationModel";
+import { PublicationModel } from "PublicationModel";
```
Comment (maryseo070) [inline packages/fan-fix-data-model/firebase/FirebaseSubscriptionModel.ts:7]: needs ../

File: packages/fan-fix-data-model/firebase/FirebaseSubscriptionModel.ts [hunk 7]
```diff
@@ -1,23 +1,40 @@
...
+  }
+  if (price) {
+    data.price = price;
+  }
+  if (publicationId) {
+    data.publicationId = publicationId;
+  }
+  if (startDate) {
+    data.startDate = startDate;
+  }
```
Comment (maryseo070) [inline packages/fan-fix-data-model/firebase/FirebaseSubscriptionModel.ts:28]: remove conditions for required ones.

File: packages/fan-fix-data-model/Subscription.ts [hunk 8]
```diff
@@ -1,10 +1,12 @@
 import { Identifiable, IdentifiableReference } from "@not-the-droids/exco-data-model";
 import { AppUser } from "./AppUser";
-import { Category } from "./Category";
+import { Publication } from "./Publication";
 
 export interface Subscription extends Identifiable {
-  creatorRef: IdentifiableReference<AppUser>;
+  appUserRef: IdentifiableReference<AppUser>;
+  publicationRef: IdentifiableReference<Publication>,
```
Comment (mattryavec) [inline packages/fan-fix-data-model/Subscription.ts:7]: Nit: semicolon

File: packages/fan-fix-data-model/SubscriptionModel.ts [hunk 9]
```diff
@@ -2,9 +2,17 @@ import { IdentifiableReference, List } from "@not-the-droids/exco-data-model";
 import { Inject } from "@not-the-droids/exco-ts-inject";
 import { Subscription } from "./Subscription";
 
+export type UpdateSubscriptionParams = Subscription;
```
Comment (mattryavec) [inline packages/fan-fix-data-model/SubscriptionModel.ts:5]: Not used yet, so delete? Just make create params `Omit<Subscription, "id">` ?

**Review by mattryavec — APPROVED** [review]
LG assuming Mary's comments are addressed

File: packages/fan-fix-data-model/PublicationModel.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+export type CreatePublicationParams = Omit<UpdatePublicationParams, "id">;
+
+export abstract class PublicationModel {
+  static inject?: Inject<PublicationModel>;
+
+  public abstract getPublicationRef(id: string): IdentifiableReference<Publication>;
+  public abstract getPublicationListByCreator(creatorId: string): List<Publication>;
+  public abstract createPublication(params: CreatePublicationParams): Promise<void>;
+  public abstract updatePublication(params: UpdatePublicationParams): Promise<void>;
```
[the remark this answers is not in the document]
Comment (helloeddiepo) [reply packages/fan-fix-data-model/PublicationModel.ts:15]: LOL! You WOULD!!!

**Pull request merged**